File: src/omnia/data/structure/rx/ObservableSet.java
package omnia.data.structure.rx;

import io.reactivex.Flowable;
import io.reactivex.functions.Function;
import omnia.data.structure.Set;

public interface ObservableSet<E> extends ObservableDataStructure {
  @Override
  ObservableChannels<E> observe();

  interface ObservableChannels<E> extends ObservableDataStructure.ObservableChannels {

    @Override
    Flowable<? extends Set<E>> states();

    @Override
    Flowable<? extends ObservableSet.MutationEvent<E>> mutations();
  }

  interface MutationEvent<E> extends ObservableDataStructure.MutationEvent {

    @Override
    Set<E> state();

    @Override
    Set<? extends ObservableSet.SetOperation<E>> operations();
  }

  @SuppressWarnings("unused")
  interface SetOperation<E> {

    static <E> Function<ObservableSet.SetOperation<E>, Flowable<ObservableSet.AddToSet<E>>>
        justAddToSetOperations() {
      return mutation -> mutation instanceof AddToSet<?>
          ? Flowable.just((AddToSet<E>) mutation)
          : Flowable.empty();
    }

    static <E> Function<ObservableSet.SetOperation<E>, Flowable<ObservableSet.RemoveFromSet<E>>>
        justRemoveFromSetOperations() {
      return mutation -> mutation instanceof RemoveFromSet<?>
          ? Flowable.just((RemoveFromSet<E>) mutation)
          : Flowable.empty();
    }
  }

  interface AddToSet<E> extends SetOperation<E> {
    E item();
  }

  interface RemoveFromSet<E> extends SetOperation<E> {
    E item();
  }
}

File: omnia/src/commonMain/omnia/data/structure/observable/ObservableSet.kt
package omnia.data.structure.observable

import com.badoo.reaktive.observable.Observable
import com.badoo.reaktive.observable.mapNotNull
import omnia.data.structure.Set

interface ObservableSet<E : Any> : ObservableDataStructure, Set<E> {

  override val observables: Observables<E>

  interface Observables<E : Any> : ObservableDataStructure.Observables {

    override val states: Observable<Set<E>>

    override val mutations: Observable<MutationEvent<E>>
  }

  interface MutationEvent<E : Any> : ObservableDataStructure.MutationEvent {

    override val state: Set<E>

    override val operations: Set<out SetOperation<E>>
  }

  interface SetOperation<E : Any> {

    val item: E

    companion object {

      fun <E : Any> justAddToSetOperations(operations: Observable<SetOperation<E>>):
          Observable<AddToSet<E>> {
        return operations.mapNotNull { it as AddToSet<E> }
      }

      fun <E : Any> justRemoveFromSetOperations(operations: Observable<SetOperation<E>>):
          Observable<RemoveFromSet<E>> {
        return operations.mapNotNull { it as RemoveFromSet<E> }
      }
    }
  }

  interface AddToSet<E : Any> : SetOperation<E>

  interface RemoveFromSet<E : Any> : SetOperation<E>
}